File: blue-core/src/main/java/blue/soundObject/SoundObjectUtilities.java
/*
 * blue - object composition environment for csound Copyright (c) 2000-2004
 * Steven Yi (stevenyi@gmail.com)
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License or (at your option) any
 * later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB. If not, write to the Free
 * Software Foundation Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307
 * USA
 */
package blue.soundObject;

import blue.noteProcessor.NoteProcessorChain;
import blue.time.TimeDuration;
import blue.time.TimePosition;
import electric.xml.Element;
import java.awt.Color;

/**
 * @author steven
 *
 * Utility class for SoundObjects to use
 */
public class SoundObjectUtilities {

    public static Element getBasicXML(SoundObject sObj) {
        Element retVal = new Element("soundObject");
        retVal.setAttribute("type", sObj.getClass().getName());

        // Save start time as TimePosition and duration as TimeDuration
        retVal.addElement(sObj.getStartTime().saveAsXML().setName("startTimePosition"));
        retVal.addElement(sObj.getSubjectiveDuration().saveAsXML().setName("subjectiveDurationTD"));
        
        retVal.addElement("name").setText(sObj.getName());

        String colorStr = Integer.toString(sObj.getBackgroundColor().getRGB());

        retVal.addElement("backgroundColor").setText(colorStr);

        if (sObj.getTimeBehavior() != TimeBehavior.NOT_SUPPORTED) {
            retVal.addElement("timeBehavior").setText(
                    Integer.toString(sObj.getTimeBehavior().getType()));
            retVal.addElement("repeatPoint").setText(
                    Double.toString(sObj.getRepeatPoint()));
        }

        if (sObj.getNoteProcessorChain() != null) {
            retVal.addElement(sObj.getNoteProcessorChain().saveAsXML());
        }

        return retVal;
    }

    public static void initBasicFromXML(Element data, SoundObject sObj)
            throws Exception {
        
        // Migration: Try new TimePosition format first, fall back to old double format
        Element startTimeUnitElement = data.getElement("startTimePosition");
        Element durationUnitElement = data.getElement("subjectiveDurationUnit");
        
        // Load start time
        if (startTimeUnitElement != null) {
            // New format: TimePosition (element is directly the timePosition)
            sObj.setStartTime(TimePosition.loadFromXML(startTimeUnitElement));
        } else if (data.getElement("startTime") != null) {
            // Old format: double (migrate to BeatTime)
            double startTime = Double.parseDouble(data.getTextString("startTime"));
            sObj.setStartTime(TimePosition.beats(startTime));
        } else {
            throw new Exception("Missing both startTimePosition and startTime elements");
        }
        
        // Load duration
        Element durationTDElement = data.getElement("subjectiveDurationTD");
        if (durationTDElement != null) {
            // New format: TimeDuration
            sObj.setSubjectiveDuration(TimeDuration.loadFromXML(durationTDElement));
        } else if (durationUnitElement != null) {
            // Legacy format: TimePosition (migrate to TimeDuration via beats)
            TimePosition legacyDur = TimePosition.loadFromXML(durationUnitElement);
            if (legacyDur instanceof TimePosition.BeatTime bt) {
                sObj.setSubjectiveDuration(TimeDuration.beats(bt.getCsoundBeats()));
            } else {
                // For non-BeatTime legacy durations, store as beats using a default context
                // This is a best-effort migration - precise conversion requires TimeContext
                sObj.setSubjectiveDuration(TimeDuration.beats(4.0));
            }
        } else if (data.getElement("subjectiveDuration") != null) {
            // Old format: double (migrate to DurationBeats)
            double duration = Double.parseDouble(data.getTextString("subjectiveDuration"));
            sObj.setSubjectiveDuration(TimeDuration.beats(duration));
        } else {
            throw new Exception("Missing subjectiveDurationTD, subjectiveDurationUnit, and subjectiveDuration elements");
        }

        String name = data.getTextString("name");

        if (name == null) {
            sObj.setName("");
        } else {
            sObj.setName(name);
        }

        if (data.getElement("backgroundColor") != null) {
            String colorStr = data.getTextString("backgroundColor");

            sObj.setBackgroundColor(new Color(Integer.parseInt(colorStr)));
        }

        if (data.getElement("timeBehavior") != null) {
            int type = Integer.parseInt(data
                    .getTextString("timeBehavior"));
            sObj.setTimeBehavior(TimeBehavior.valueByType(type));
        }

        if (data.getElement("repeatPoint") != null) {
            sObj.setRepeatPoint(Double.parseDouble(data
                    .getTextString("repeatPoint")));
        }

        if (data.getElement("noteProcessorChain") != null) {
            sObj.setNoteProcessorChain(NoteProcessorChain.loadFromXML(data
                    .getElement("noteProcessorChain")));
        }
    }

    public static boolean isOrContainsInstance(SoundObject sObj) {
        if (sObj instanceof Instance) {
            return true;
        }

        if (sObj instanceof PolyObject pObj) {
            for (SoundObject soundObject : pObj.getSoundObjects(true)) {
                if (isOrContainsInstance(soundObject)) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: packages/blue-data/src/sound-objects/sound-object-utilities.ts
/**
 * SoundObjectUtilities — shared utility for loading common sound object properties from XML.
 * Mirrors the Java SoundObjectUtilities.initBasicFromXML.
 *
 * Handles 3 XML formats for startTime and subjectiveDuration:
 *   1. New format: `<startTime type='BEATS'><csoundBeats>8.0</csoundBeats></startTime>`
 *   2. Legacy tag: `<startTimePosition type='BEATS'>...</startTimePosition>`
 *   3. Old format: `<startTime>8.0</startTime>` (plain text, no type attr)
 */
import { TimePosition } from '../time/time-position';
import { TimeDuration } from '../time/time-duration';
import { TimeBehavior } from './time-behavior';
import { NoteProcessorChain } from '../note-processors/note-processor-chain';
import { Element } from '../serialization/xml-reader';

/**
 * Interface for objects that have the basic sound object properties.
 * Both AbstractSoundObject and PolyObject implement this.
 */
export interface BasicSoundObject {
  setName(name: string): void;
  getName(): string;
  setStartTime(value: TimePosition): void;
  getStartTime(): TimePosition;
  setSubjectiveDuration(value: TimeDuration): void;
  getSubjectiveDuration(): TimeDuration;
  setTimeBehavior(behavior: TimeBehavior): void;
  getTimeBehavior(): TimeBehavior;
  setBackgroundColor(color: number): void;
  getBackgroundColor(): number;
  setRepeatPoint(rp: TimeDuration | null): void;
  getRepeatPoint(): TimeDuration | null;
  setNoteProcessorChain(chain: NoteProcessorChain): void;
  getNoteProcessorChain(): NoteProcessorChain;
}

/**
 * Load common sound object properties from XML, handling all 3 format variants
 * for startTime and subjectiveDuration.
 *
 * @param sObj The sound object to populate
 * @param data The XML element containing the sound object data
 */
export function initBasicFromXML(sObj: BasicSoundObject, data: Element): void {
  // ─── Name ───
  const name = data.getTextString('name');
  if (name) sObj.setName(name);

  // ─── Start Time (3 formats) ───
  const startTimeElement = data.getElement('startTime');
  if (startTimeElement) {
    const typeAttr = startTimeElement.getAttributeValue('type');
    if (typeAttr) {
      // Format 1: New nested format with type attribute
      sObj.setStartTime(TimePosition.loadFromXML(startTimeElement));
    } else {
      // Format 3: Old plain text format
      const text = startTimeElement.getTextString();
      if (text) {
        sObj.setStartTime(TimePosition.beats(parseFloat(text)));
      }
    }
  } else {
    // Format 2: Legacy startTimePosition tag
    const startPosElement = data.getElement('startTimePosition');
    if (startPosElement) {
      const typeAttr = startPosElement.getAttributeValue('type');
      if (typeAttr) {
        sObj.setStartTime(TimePosition.loadFromXML(startPosElement));
      } else {
        const text = startPosElement.getTextString();
        if (text) {
          sObj.setStartTime(TimePosition.beats(parseFloat(text)));
        }
      }
    }
  }

  // ─── Subjective Duration (3 formats) ───
  const durElement = data.getElement('subjectiveDuration');
  if (durElement) {
    const typeAttr = durElement.getAttributeValue('type');
    if (typeAttr) {
      // Format 1: New nested format
      sObj.setSubjectiveDuration(TimeDuration.loadFromXML(durElement));
    } else {
      // Format 3: Old plain text
      const text = durElement.getTextString();
      if (text) {
        sObj.setSubjectiveDuration(TimeDuration.beats(parseFloat(text)));
      }
    }
  } else {
    // Format 2: Legacy subjectiveDurationTD tag
    const durTDElement = data.getElement('subjectiveDurationTD');
    if (durTDElement) {
      const typeAttr = durTDElement.getAttributeValue('type');
      if (typeAttr) {
        sObj.setSubjectiveDuration(TimeDuration.loadFromXML(durTDElement));
      } else {
        const text = durTDElement.getTextString();
        if (text) {
          sObj.setSubjectiveDuration(TimeDuration.beats(parseFloat(text)));
        }
      }
    } else {
      // Legacy: subjectiveDurationUnit (TimePosition used as duration)
      const durUnitElement = data.getElement('subjectiveDurationUnit');
      if (durUnitElement) {
        const typeAttr = durUnitElement.getAttributeValue('type');
        if (typeAttr) {
          // Load as TimePosition, extract beat value, create Duration
          const pos = TimePosition.loadFromXML(durUnitElement);
          sObj.setSubjectiveDuration(TimeDuration.beats(pos.getValue()));
        } else {
          const text = durUnitElement.getTextString();
          if (text) {
            sObj.setSubjectiveDuration(TimeDuration.beats(parseFloat(text)));
          }
        }
      }
    }
  }

  // ─── Time Behavior ───
  // Java uses ordinals: NOT_SUPPORTED=-1, SCALE=0, REPEAT_CLASSIC=1, NONE=2, REPEAT=3
  const tbStr = data.getTextString('timeBehavior');
  if (tbStr) {
    if (Object.values(TimeBehavior).includes(tbStr as TimeBehavior)) {
      sObj.setTimeBehavior(tbStr as TimeBehavior);
    } else {
      // Legacy numeric format
      const tbNum = parseInt(tbStr, 10);
      switch (tbNum) {
        case -1:
          sObj.setTimeBehavior(TimeBehavior.NOT_SUPPORTED);
          break;
        case 0:
          sObj.setTimeBehavior(TimeBehavior.SCALE);
          break;
        case 1:
          sObj.setTimeBehavior(TimeBehavior.REPEAT_CLASSIC);
          break;
        case 3:
          sObj.setTimeBehavior(TimeBehavior.REPEAT);
          break;
        case 2:
          sObj.setTimeBehavior(TimeBehavior.NONE);
          break;
      }
    }
  }

  // ─── Background Color ───
  const colorStr = data.getTextString('backgroundColor');
  if (colorStr) {
    sObj.setBackgroundColor(parseInt(colorStr, 10));
  }

  // ─── Repeat Point ───
  // Java: only set repeatPoint if value > 0.0; -1 means "no repeat" → null
  const rpElement = data.getElement('repeatPoint');
  if (rpElement) {
    const typeAttr = rpElement.getAttributeValue('type');
    if (typeAttr) {
      sObj.setRepeatPoint(TimeDuration.loadFromXML(rpElement));
    } else {
      const text = rpElement.getTextString();
      if (text) {
        const rpVal = parseFloat(text);
        if (rpVal > 0.0) {
          sObj.setRepeatPoint(TimeDuration.beats(rpVal));
        } else {
          sObj.setRepeatPoint(null);
        }
      }
    }
  }

  // ─── Note Processor Chain ───
  const npcElement = data.getElement('noteProcessorChain');
  if (npcElement) {
    sObj.setNoteProcessorChain(NoteProcessorChain.loadFromXML(npcElement));
  }
}

function timeBehaviorToType(tb: TimeBehavior): number {
  switch (tb) {
    case TimeBehavior.SCALE:
      return 0;
    case TimeBehavior.REPEAT_CLASSIC:
      return 1;
    case TimeBehavior.NONE:
      return 2;
    case TimeBehavior.REPEAT:
      return 3;
    case TimeBehavior.NOT_SUPPORTED:
      return -1;
    default:
      return 0;
  }
}

export function getBasicXML(sObj: BasicSoundObject, javaType: string): Element {
  const elem = new Element('soundObject');
  elem.setAttribute('type', javaType);

  elem.addElement(sObj.getStartTime().saveAsXML().setName('startTime'));
  elem.addElement(sObj.getSubjectiveDuration().saveAsXML().setName('subjectiveDuration'));
  elem.addElement('name').setText(sObj.getName());
  elem.addElement('backgroundColor').setText(sObj.getBackgroundColor().toString());

  const tb = sObj.getTimeBehavior();
  if (tb !== TimeBehavior.NOT_SUPPORTED) {
    elem.addElement('timeBehavior').setText(timeBehaviorToType(tb).toString());
    const rp = sObj.getRepeatPoint();
    if (rp) {
      elem.addElement(rp.saveAsXML().setName('repeatPoint'));
    } else {
      elem.addElement('repeatPoint').setText('-1.0');
    }
  }

  const npc = sObj.getNoteProcessorChain();
  if (npc) {
    elem.addElement(npc.saveAsXML().setName('noteProcessorChain'));
  }

  return elem;
}
